src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngAlgorithmCore.cs

```diff
@@ -18,13 +18,35 @@ public static CngKey Duplicate(CngKey key)
             return CngKey.Open(key.Handle, key.IsEphemeral ? CngKeyHandleOpenOptions.EphemeralKey : CngKeyHandleOpenOptions.None);
         }
 
-        public CngKey GetOrGenerateKey(int keySize, CngAlgorithm algorithm)
+        public CngKey GetOrGenerateKey(int keySize, CngAlgorithm algorithm, bool forceNewKey = false)
         {
-            // If our key size was changed from the key we're using, we need to generate a new key.
-            if (_lazyKey != null && _lazyKey.KeySize != keySize)
+            if (algorithm == null)
+                throw new ArgumentNullException(nameof(algorithm));
+
+            if (_lazyKey != null)
             {
-                _lazyKey.Dispose();
-                _lazyKey = null;
+                if (_lazyKey.IsECNamedCurve())
+                {
+                    if (forceNewKey)
+                    {
+                        string curveName = _lazyKey.GetCurveName();
+                        if (!string.IsNullOrEmpty(curveName))
+                        {
+                            algorithm = CngAlgorithm.ECDsa(new ECNamedCurve(new Oid(null, curveName)));
```

Comment:
This is part of the work mentioned earlier that is needed to have a consistent way to create an Oid from either the oid value or friendly name (try oid value first if present, then try friendly name)